src/serac/numerics/assembled_sparse_matrix.cpp

```diff
@@ -0,0 +1,297 @@
+
+#include "mfem/linalg/dtensor.hpp"
+#include "assembled_sparse_matrix.hpp"
+
+namespace detail {
+/**
+ * We'd like to access the protected offset, indices, and gatherMap variables from mfem::ElementRestriction
+ * to compute our sparsity patterns in a manner consistent with MFEM.
+ * In order to do that, we use the following template tricks.
+ * Here's an article that explains the trick being used: https://accu.org/journals/overload/28/156/harrison_2776/
+ */
+template <auto offsetV, auto indicesV, auto gatherV>
+struct forbidden_restriction {
+  friend mfem::Array<int>& ElementRestriction_offsets(mfem::ElementRestriction& from) { return from.*offsetV; }
```

Comment:
Minor: we normally use PascalCase for struct names (with the exception being type traits) and camelCase for function names - for the methods here they're referring to the MFEM type so maybe `ElementRestrictionOffsets` etc?